events/events.go

```diff
@@ -0,0 +1,197 @@
+	Get() interface{}
+}
+
+// Publisher is a publisher for an event
+type Publisher interface {
+	Event
+
+	// GetAndSubscribe returns the latest Event and a subscriber
+	GetAndSubscribe() (interface{}, Subscriber, error)
+	// NotifyAll notifies all subscribers
+	NotifyAll() error
+	// SubscriberLen returns the number of subscribers
+	SubscriberLen() int
+	// Close shuts down the publisher
+	Close()
+}
+
+// NewPublisher returns a new event publisher
+func NewPublisher() Publisher {
+	return &publisher{}
+}
+
+// Subscriber is a subscriber to an event
+type Subscriber interface {
+	Event
+
+	// Notifications returns the channel of notifications
+	Notifications() <-chan struct{}
+	// OnClose registers a function to be called on Close()
+	OnClose(func())
+	// Close closes the subscriber
+	Close()
+}
+
+// NewSubscriber returns a new event subscriber
+func NewSubscriber(ch <-chan struct{}) Subscriber {
+	return &subscriber{notificationCh: ch}
+}
+
+type publisher struct {
+	sync.RWMutex
+
+	onGetFunc   GetFn
+	subscribers []chan struct{}
+	closed      bool
+}
+
+func (p *publisher) OnGet(f GetFn) {
+	p.Lock()
+	p.onGetFunc = f
+	p.Unlock()
+}
+
+func (p *publisher) Get() interface{} {
+	p.RLock()
+	defer p.RUnlock()
+	if p.onGetFunc == nil {
+		return nil
+	}
+	return p.onGetFunc()
+}
+
+func (p *publisher) GetAndSubscribe() (interface{}, Subscriber, error) {
+	c := make(chan struct{}, 2)
+	closeFn, err := p.registerChannel(c)
+	if err != nil {
+		return p.Get(), nil, err
+	}
+	s := NewSubscriber(c)
+	s.OnGet(p.Get)
+	s.OnClose(closeFn)
+	return p.Get(), s, nil
+}
+
+func (p *publisher) NotifyAll() error {
+	p.RLock()
+	defer p.RUnlock()
+	if p.closed {
+		return errClosed
+	}
+	for _, s := range p.subscribers {
+		select {
+		case s <- struct{}{}:
+		}
+	}
+	return nil
+}
+
+func (p *publisher) SubscriberLen() int {
+	return len(p.subscribers)
+}
+
+func (p *publisher) Close() {
+	p.Lock()
+	defer p.Unlock()
+	if p.closed {
+		return
+	}
+	p.closed = true
+	for _, s := range p.subscribers {
+		close(s)
+	}
+	p.subscribers = p.subscribers[:0]
+}
+
+func (p *publisher) registerChannel(c chan struct{}) (func(), error) {
+	p.Lock()
+	defer p.Unlock()
+	if p.closed {
+		return nil, errClosed
+	}
+	p.subscribers = append(p.subscribers, c)
+	return p.closeFunc(c), nil
+}
+
+func (p *publisher) closeFunc(c chan struct{}) func() {
+	return func() {
+		p.Lock()
+		defer p.Unlock()
+		if p.closed {
+			return
+		}
+
+		close(c)
+
+		for i, s := range p.subscribers {
+			if s == c {
+				p.subscribers = append(p.subscribers[:i], p.subscribers[i+1:]...)
+				return
+			}
+		}
+	}
+}
+
+type subscriber struct {
+	sync.RWMutex
+
+	notificationCh <-chan struct{}
+	onGetFn        GetFn
+	onCloseFn      func()
+	closed         bool
+}
+
+func (s *subscriber) Notifications() <-chan struct{} {
+	return s.notificationCh
+}
+
+func (s *subscriber) OnGet(f GetFn) {
+	s.Lock()
+	s.onGetFn = f
+	s.Unlock()
+}
+
+func (s *subscriber) Get() interface{} {
```

Comment:
Yeah my plan was for whoever creates the publisher to trigger the notification rather than the publisher to watch for updates.